link-grammar/fast-match.c

```diff
@@ -203,6 +204,59 @@ static Match_node * add_to_left_table_list(Match_node * m, Match_node * l)
 }
 
 /**
+ * Validate that the uppercase part of two connectors are the same.
+ * Return true if so, false otherwise.
+ * FIXME: Use connector enumeration.
+ */
+static bool con_uc_eq(const Connector *c1, const Connector *c2)
+{
+
+	if (string_set_cmp(c1->string, c2->string)) return true;
+	if (c1->hash != c2->hash) return false;
+	if (c1->uc_length != c1->uc_length) return false;
+
+	const char *uc1 = &c1->string[c1->uc_start];
+	const char *uc2 = &c2->string[c2->uc_start];
+	/* We arrive here for less than 50% of the cases for "en" and
+	 * less then 20% of the cases for "ru", and the following
+	 * condition is always true.
+	 * How come 2 different uc strings never have the same hash value? */
```

Comment:
I mentioned "slots" in my message above, but of course this doesn't have anything to do with the number of slots in the hash table, since I compare the raw hash values, not after masking them with the table size.
So different hash values of several K of strings over 64K values is actually making sense...
And it seems that in any case the check is needed since there is no guarantee that an added different UC will not get an already-used value.

I will wait to hear all your comment about this proposed speedup and then will add commits as needed,